beacon_node/http_api/src/publish_blocks.rs

```diff
@@ -119,130 +144,252 @@ pub async fn publish_block<T: BeaconChainTypes, B: IntoGossipVerifiedBlockConten
-    let (gossip_verified_block, gossip_verified_blobs) =
-        match block_contents.into_gossip_verified_block(&chain) {
-            Ok(b) => b,
-            Err(BlockContentsError::BlockError(BlockError::BlockIsAlreadyKnown(_)))
-            | Err(BlockContentsError::BlobError(
-                beacon_chain::blob_verification::GossipBlobError::RepeatBlob { .. },
-            )) => {
-                // Allow the status code for duplicate blocks to be overridden based on config.
-                return Ok(warp::reply::with_status(
-                    warp::reply::json(&ErrorMessage {
-                        code: duplicate_status_code.as_u16(),
-                        message: "duplicate block".to_string(),
-                        stacktraces: vec![],
-                    }),
-                    duplicate_status_code,
-                )
-                .into_response());
-            }
-            Err(e) => {
-                warn!(
-                    log,
-                    "Not publishing block - not gossip verified";
-                    "slot" => slot,
-                    "error" => %e
-                );
-                return Err(warp_utils::reject::custom_bad_request(e.to_string()));
-            }
-        };
 
-    // Clone here, so we can take advantage of the `Arc`. The block in `BlockContents` is not,
-    // `Arc`'d but blobs are.
-    let block = gossip_verified_block.block.block_cloned();
-    let blobs_opt = gossip_verified_blobs.as_ref().map(|gossip_verified_blobs| {
-        let blobs = gossip_verified_blobs
+    // Check blob inclusion proofs and convert to blob sidecars ready for publication.
+    let mut blob_sidecars = if let Some((kzg_proofs, blobs)) = unverified_blobs {
+        kzg_proofs
             .into_iter()
-            .map(|b| b.clone_blob())
-            .collect::<Vec<_>>();
-        VariableList::from(blobs)
-    });
+            .zip(blobs)
+            .enumerate()
+            .map(|(i, (proof, unverified_blob))| {
+                let _timer = metrics::start_timer(
+                    &beacon_chain::metrics::BLOB_SIDECAR_INCLUSION_PROOF_COMPUTATION,
+                );
+                let blob_sidecar =
+                    BlobSidecar::new(i, unverified_blob, &block, proof).map(Arc::new);
+                blob_sidecar.map_err(|e| {
+                    error!(
+                        log,
+                        "Invalid blob - not publishing block";
+                        "error" => ?e,
+                        "blob_index" => i,
+                        "slot" => slot,
+                    );
+                    warp_utils::reject::custom_bad_request(format!("{e:?}"))
+                })
+            })
+            .collect::<Result<Vec<_>, Rejection>>()?
+    } else {
+        vec![]
+    };
 
-    let block_root = block_root.unwrap_or(gossip_verified_block.block_root);
+    // Gossip verify the block and blobs separately.
```

Comment:
I think we'll have to break down some of the logic here into smaller functions - this will be really long when we introduce peerdas, but I'm happy to do this later in a separate PR.